File: apps/dashboard/src/types/api/training.ts
import { z } from 'zod';

/**
 * Wire schemas for the Machine Learning Training Framework API (see
 * `services/api/app/api/v1/endpoints/training.py` /
 * `services/api/app/schemas/training.py`). Every response is validated
 * against these before the app trusts it, matching `experiments.ts`'s
 * own convention.
 */

export const TRAINING_JOB_STATUSES = [
  'pending',
  'running',
  'completed',
  'failed',
  'cancelled',
] as const;

export const TrainingJobStatusSchema = z.enum(TRAINING_JOB_STATUSES);

export type TrainingJobStatus = z.infer<typeof TrainingJobStatusSchema>;

export const TRAINING_JOB_STAGES = [
  'validate_dataset',
  'load_dataset',
  'initialize_model',
  'execute_training',
  'save_results',
  'update_experiment',
] as const;

export const TrainingJobStageSchema = z.enum(TRAINING_JOB_STAGES);

export type TrainingJobStage = z.infer<typeof TrainingJobStageSchema>;

export const TRAINING_LOG_LEVELS = ['debug', 'info', 'warning', 'error'] as const;

export const TrainingLogLevelSchema = z.enum(TRAINING_LOG_LEVELS);

export type TrainingLogLevel = z.infer<typeof TrainingLogLevelSchema>;

export const TrainingJobLogSchema = z.object({
  id: z.string(),
  level: TrainingLogLevelSchema,
  stage: TrainingJobStageSchema.nullable(),
  message: z.string(),
  logged_at: z.string().datetime(),
});

export type TrainingJobLog = z.infer<typeof TrainingJobLogSchema>;

export const TrainingJobSchema = z.object({
  id: z.string(),
  experiment_id: z.string(),
  dataset_version: z.string().nullable(),
  symbol: z.string().nullable(),
  timeframe: z.string().nullable(),
  target_column: z.string().nullable(),
  model_type: z.string(),
  hyperparameters: z.record(z.string(), z.unknown()),
  /** Whether numeric feature columns were (or will be) z-score normalized —
   * fit on the train split alone — before this job's model trains/predicts.
   * Ignored by an adapter whose requires_real_data is false. */
  normalize_features: z.boolean(),
  status: TrainingJobStatusSchema,
  current_stage: TrainingJobStageSchema.nullable(),
  error_message: z.string().nullable(),
  error_detail: z
    .object({
      reason: z.string(),
      affected_feature: z.string().nullable(),
      affected_rows: z.array(z.number()).nullable(),
      suggested_fix: z.string(),
    })
    .nullable()
    .optional(),
  result_summary: z.record(z.string(), z.unknown()).nullable(),
  started_at: z.string().datetime().nullable(),
  completed_at: z.string().datetime().nullable(),
  logs: z.array(TrainingJobLogSchema),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
});

export type TrainingJob = z.infer<typeof TrainingJobSchema>;

export const TrainingJobSummarySchema = z.object({
  id: z.string(),
  experiment_id: z.string(),
  dataset_version: z.string().nullable(),
  model_type: z.string(),
  status: TrainingJobStatusSchema,
  current_stage: TrainingJobStageSchema.nullable(),
  log_count: z.number().int().nonnegative(),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
});

export type TrainingJobSummary = z.infer<typeof TrainingJobSummarySchema>;

export const TrainingJobListResponseSchema = z.object({
  jobs: z.array(TrainingJobSummarySchema),
  total: z.number().int().nonnegative(),
  limit: z.number().int().positive(),
  offset: z.number().int().nonnegative(),
  statuses: z.array(z.string()),
  stages: z.array(z.string()),
});

export type TrainingJobListResponse = z.infer<typeof TrainingJobListResponseSchema>;

export const MODEL_KINDS = ['placeholder', 'classification', 'regression'] as const;

export const ModelKindSchema = z.enum(MODEL_KINDS);

export type ModelKind = z.infer<typeof ModelKindSchema>;

export const ModelAdapterSchema = z.object({
  name: z.string(),
  label: z.string(),
  description: z.string(),
  framework: z.string(),
  model_kind: ModelKindSchema,
  requires_real_data: z.boolean(),
  hyperparameter_hints: z.array(z.string()),
  version: z.string(),
});

export type ModelAdapter = z.infer<typeof ModelAdapterSchema>;

export const ModelAdapterCatalogResponseSchema = z.object({
  adapters: z.array(ModelAdapterSchema),
});

export type ModelAdapterCatalogResponse = z.infer<typeof ModelAdapterCatalogResponseSchema>;

export const TrainingArtifactSchema = z.object({
  artifact_type: z.string(),
  filename: z.string(),
  content_type: z.string(),
  download_url: z.string(),
});

export type TrainingArtifact = z.infer<typeof TrainingArtifactSchema>;

export const TrainingArtifactListResponseSchema = z.object({
  job_id: z.string(),
  artifacts: z.array(TrainingArtifactSchema),
});

export type TrainingArtifactListResponse = z.infer<typeof TrainingArtifactListResponseSchema>;

export const TrainingJobPredictResponseSchema = z.object({
  predictions: z.array(z.unknown()),
  feature_columns: z.array(z.string()).nullable().optional(),
  classes: z.array(z.unknown()).nullable().optional(),
  probabilities: z.array(z.array(z.number())).nullable().optional(),
  confidence_levels: z.array(z.string()).nullable().optional(),
});

export type TrainingJobPredictResponse = z.infer<typeof TrainingJobPredictResponseSchema>;

File: services/api/app/schemas/training.py
"""Wire schemas (DTOs) for the Machine Learning Training Framework API.

The ORM models (`app/models/training.py`) stay the source of truth for
persisted shape; this module is the one place a create request is
validated and a response is assembled — the same split
`app/schemas/experiments.py` already follows.
"""

import mimetypes
import uuid
from datetime import UTC, datetime
from pathlib import Path
from typing import TYPE_CHECKING, Any, Literal

from pydantic import BaseModel, Field, field_serializer

from app.models.training import TRAINING_JOB_STAGES, TRAINING_JOB_STATUSES, TRAINING_LOG_LEVELS
from app.training.base import ModelAdapterMetadata, ModelKind

if TYPE_CHECKING:
    from app.models.training import TrainingJob, TrainingJobLog

TrainingJobStatus = Literal["pending", "running", "completed", "failed", "cancelled"]
TrainingJobStage = Literal[
    "validate_dataset",
    "load_dataset",
    "initialize_model",
    "execute_training",
    "save_results",
    "update_experiment",
]
TrainingLogLevel = Literal["debug", "info", "warning", "error"]


def _iso(value: datetime) -> str:
    """ISO-8601 UTC with a literal `Z`, matching every timestamp on this API."""
    return value.astimezone(UTC).isoformat().replace("+00:00", "Z")


class TrainingJobCreateRequest(BaseModel):
    """A new training job registration, linked to an existing experiment."""

    experiment_id: uuid.UUID = Field(..., description="The experiment this job trains for")
    model_type: str = Field(
        ...,
        min_length=1,
        max_length=100,
        description="A registered model adapter name — see GET /training-jobs/models",
    )
    dataset_version: str | None = Field(
        default=None,
        max_length=200,
        description="Overrides the experiment's own dataset_version; defaults from it if omitted",
    )
    symbol: str | None = Field(
        default=None,
        max_length=50,
        description=(
            "Market symbol to load real candles from — required if the chosen model "
            "adapter's requires_real_data is true"
        ),
    )
    timeframe: str | None = Field(
        default=None,
        max_length=20,
        description="Candle timeframe to load, e.g. '1h' — required alongside symbol",
    )
    target_column: str | None = Field(
        default=None,
        max_length=100,
        description="Which built target column to predict; defaults to the first one built",
    )
    hyperparameters: dict[str, Any] = Field(default_factory=dict)


class TrainingJobLogDTO(BaseModel):
    id: str
    level: TrainingLogLevel
    stage: TrainingJobStage | None
    message: str
    logged_at: datetime

    @field_serializer("logged_at")
    def _serialize_logged_at(self, value: datetime) -> str:
        return _iso(value)

    @classmethod
    def from_model(cls, log: "TrainingJobLog") -> "TrainingJobLogDTO":
        return cls(
            id=str(log.id),
            level=log.level,  # type: ignore[arg-type]
            stage=log.stage,  # type: ignore[arg-type]
            message=log.message,
            logged_at=log.logged_at,
        )


class TrainingJobResponse(BaseModel):
    """A full training job record, including its logs."""

    id: str
    experiment_id: str
    dataset_version: str | None
    symbol: str | None
    timeframe: str | None
    target_column: str | None
    model_type: str
    hyperparameters: dict[str, Any]
    status: TrainingJobStatus
    current_stage: TrainingJobStage | None
    error_message: str | None
    error_detail: dict[str, Any] | None = Field(
        default=None,
        description=(
            "A structured failure report (reason, affected_feature, affected_rows, "
            "suggested_fix) recorded alongside error_message when a run fails."
        ),
    )
    result_summary: dict[str, Any] | None
    started_at: datetime | None
    completed_at: datetime | None
    logs: list[TrainingJobLogDTO]
    created_at: datetime
    updated_at: datetime

    @field_serializer("started_at", "completed_at", "created_at", "updated_at")
    def _serialize_optional_timestamps(self, value: datetime | None) -> str | None:
        return None if value is None else _iso(value)

    @classmethod
    def from_model(cls, job: "TrainingJob") -> "TrainingJobResponse":
        return cls(
            id=str(job.id),
            experiment_id=str(job.experiment_id),
            dataset_version=job.dataset_version,
            symbol=job.symbol,
            timeframe=job.timeframe,
            target_column=job.target_column,
            model_type=job.model_type,
            hyperparameters=job.hyperparameters or {},
            status=job.status,  # type: ignore[arg-type]
            current_stage=job.current_stage,  # type: ignore[arg-type]
            error_message=job.error_message,
            error_detail=job.error_detail,
            result_summary=job.result_summary,
            started_at=job.started_at,
            completed_at=job.completed_at,
            logs=[TrainingJobLogDTO.from_model(log) for log in job.logs],
            created_at=job.created_at,
            updated_at=job.updated_at,
        )


class TrainingJobSummaryDTO(BaseModel):
    """One row in the training job list — metadata only, to keep a list page light."""

    id: str
    experiment_id: str
    dataset_version: str | None
    model_type: str
    status: TrainingJobStatus
    current_stage: TrainingJobStage | None
    log_count: int
    created_at: datetime
    updated_at: datetime

    @field_serializer("created_at", "updated_at")
    def _serialize_timestamps(self, value: datetime) -> str:
        return _iso(value)

    @classmethod
    def from_model(cls, job: "TrainingJob") -> "TrainingJobSummaryDTO":
        return cls(
            id=str(job.id),
            experiment_id=str(job.experiment_id),
            dataset_version=job.dataset_version,
            model_type=job.model_type,
            status=job.status,  # type: ignore[arg-type]
            current_stage=job.current_stage,  # type: ignore[arg-type]
            log_count=len(job.logs),
            created_at=job.created_at,
            updated_at=job.updated_at,
        )


class TrainingJobListResponse(BaseModel):
    """One page of training jobs plus the metadata a list UI needs to paginate and filter."""

    jobs: list[TrainingJobSummaryDTO]
    total: int
    limit: int
    offset: int
    statuses: list[str] = Field(default_factory=lambda: list(TRAINING_JOB_STATUSES))
    stages: list[str] = Field(default_factory=lambda: list(TRAINING_JOB_STAGES))


class ModelAdapterDTO(BaseModel):
    """One entry in the model adapter catalogue — the frontend's model-type dropdown source."""

    name: str
    label: str
    description: str
    framework: str
    model_kind: ModelKind = Field(
        description=(
            "'classification' or 'regression' for a real baseline adapter, "
            "'placeholder' for the fabricated one — tells the frontend whether to "
            "render a confusion matrix or regression metrics."
        )
    )
    requires_real_data: bool = Field(
        description="Whether a job using this adapter requires symbol/timeframe/target_column"
    )
    hyperparameter_hints: list[str]
    version: str

    @classmethod
    def from_metadata(cls, metadata: ModelAdapterMetadata) -> "ModelAdapterDTO":
        return cls(
            name=metadata.name,
            label=metadata.label,
            description=metadata.description,
            framework=metadata.framework,
            model_kind=metadata.model_kind,
            requires_real_data=metadata.requires_real_data,
            hyperparameter_hints=list(metadata.hyperparameter_hints),
            version=metadata.version,
        )


class ModelAdapterCatalogResponse(BaseModel):
    """Every registered model adapter, for a frontend dropdown/catalogue view."""

    adapters: list[ModelAdapterDTO]


class TrainingJobPredictRequest(BaseModel):
    """One or more feature rows to predict for, using a completed job's trained model."""

    rows: list[list[float]] = Field(
        ...,
        min_length=1,
        description="Each row must have exactly as many values as the job's feature_columns",
    )


class TrainingJobPredictResponse(BaseModel):
    """Predictions for each requested row, in the same order.

    `probabilities`/`confidence_levels`/`classes` are populated only when the job's
    model adapter supports `predict_proba` (today: `logistic_regression`) — `None`
    for every other adapter, never a fabricated confidence.
    """

    predictions: list[Any]
    feature_columns: list[str] | None = Field(
        default=None,
        description="The feature columns (in order) this job's model expects, if recorded",
    )
    classes: list[Any] | None = Field(
        default=None,
        description="The model's class labels, in the same order as each row of probabilities",
    )
    probabilities: list[list[float]] | None = Field(
        default=None,
        description="Per-row, per-class probability, aligned with classes — only for a classifier",
    )
    confidence_levels: list[str] | None = Field(
        default=None,
        description="Per-row 'high'/'medium'/'low', from the predicted class's own probability",
    )


class TrainingArtifactDTO(BaseModel):
    """One downloadable file produced by a completed job's training run."""

    artifact_type: str = Field(description="e.g. 'model_joblib', 'metrics_json', 'roc_curve_png'")
    filename: str
    content_type: str
    download_url: str

    @classmethod
    def build(cls, job_id: uuid.UUID, artifact_type: str, uri: str) -> "TrainingArtifactDTO":
        filename = Path(uri.removeprefix("file://")).name
        content_type, _ = mimetypes.guess_type(filename)
        return cls(
            artifact_type=artifact_type,
            filename=filename,
            content_type=content_type or "application/octet-stream",
            download_url=f"/api/v1/training-jobs/{job_id}/artifacts/{artifact_type}",
        )


class TrainingArtifactListResponse(BaseModel):
    """Every downloadable artifact a completed job's training run produced."""

    job_id: str
    artifacts: list[TrainingArtifactDTO]


__all__ = [
    "TRAINING_LOG_LEVELS",
    "ModelAdapterCatalogResponse",
    "ModelAdapterDTO",
    "TrainingArtifactDTO",
    "TrainingArtifactListResponse",
    "TrainingJobCreateRequest",
    "TrainingJobListResponse",
    "TrainingJobLogDTO",
    "TrainingJobPredictRequest",
    "TrainingJobPredictResponse",
    "TrainingJobResponse",
    "TrainingJobSummaryDTO",
]
